src/util/fifo.h

```diff
@@ -67,3 +82,162 @@ class FIFO {
     PaUtilRingBuffer m_ringBuffer;
     DISALLOW_COPY_AND_ASSIGN(FIFO);
 };
+} // namespace PA
+
+// Fast, trivial type only single producer single consumer ring buffer, lock and wait free.
+// Internal buffer size will be rounded up to a power of two.
+
+// Internally we use std::size_t, but the original API used int for the return values.
+// It would be better to use std::size_t also for the return values, but this breaks
+// Windows compilation and should be done in a follow-up PR.
+// using ring_buffer_size_t = int;
+
+template<class DataType>
+class FIFO {
+  public:
+    using size_type = std::size_t;
+
+    explicit FIFO(size_type size)
+            : m_size{roundUpToPowerOf2(static_cast<unsigned int>(size))},
+              m_mask{m_size - 1},
+              m_data(m_size),
+              m_writeIndex{0},
+              m_readIndex{0} {
+    }
+
+    // Returns the number of values in the ringbuffer available for read
+    int readAvailable() const {
+        const size_type readIndex = m_readIndex.load(std::memory_order_relaxed);
+        const size_type writeIndex = m_writeIndex.load(std::memory_order_acquire);
+        return static_cast<int>(writeIndex - readIndex);
+    }
+    // Returns the space in the ringbuffer available for write
+    int writeAvailable() const {
+        const size_type readIndex = m_readIndex.load(std::memory_order_acquire);
+        const size_type writeIndex = m_writeIndex.load(std::memory_order_relaxed);
```

Comment:
Following the logic below should we swap both loads?

Reply:
No, it doesn't matter. The important thing here is that the m_readIndex is loaded with std::memory_order_acquire, as it is changed in the other thread, and we want to make sure we see changed value here. m_writeIndex will not be modified in any other thread than the current thread, so it's correct value will always be guaranteed in this thread. Swapping the two loads would make no difference.